### PR TITLE
[lldb] Implement DynamicSelf case for Swift DumpTypeValue functions

### PR DESCRIPTION
**Pull request by augusto2112**

### CODE REVIEW

File: lldb/test/API/lang/swift/designated_initializer_self/main.swift [hunk 1]
```diff
@@ -3,6 +3,7 @@ class C {
   convenience init(unused: Bool) { 
     self.init()
     print(1)//%self.expect('po self', substrs=['<C: 0x'])
+            //%self.expect('v self', substrs=['(Self) self = 0x'])
```
Comment (kastiglione) [inline lldb/test/API/lang/swift/designated_initializer_self/main.swift:6]: I feel like we shouldn't want this to show as `Self`.

File: lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp [hunk 2]
```diff
@@ -7579,6 +7579,11 @@ bool SwiftASTContext::DumpTypeValue(
 
   swift::CanType swift_can_type(GetCanonicalSwiftType(type));
 
+  if (auto *dyn_self = llvm::dyn_cast_or_null<swift::DynamicSelfType>(
+          swift_can_type.getPointer()))
```
Comment (adrian-prantl) [inline lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:7583]: Should this just be part of GetConanicalSwiftType?

File: lldb/source/Plugins/TypeSystem/Swift/TypeSystemSwiftTypeRef.cpp [hunk 3]
```diff
@@ -3418,6 +3418,15 @@ bool TypeSystemSwiftTypeRef::DumpTypeValue(
     using namespace swift::Demangle;
     Demangler dem;
     auto *node = DemangleCanonicalType(dem, type);
+    // Get the actual type of self.
+    if (node->getKind() == Node::Kind::DynamicSelf) {
```
Comment (adrian-prantl) [inline lldb/source/Plugins/TypeSystem/Swift/TypeSystemSwiftTypeRef.cpp:3422]: Same question here.

File: lldb/test/API/lang/swift/designated_initializer_self/main.swift (hunk 1 above)
Comment (adrian-prantl) [reply lldb/test/API/lang/swift/designated_initializer_self/main.swift:6]: Self is the type as it appears in the source code. Maube the static type should be Self and the dynamic type should be the actual type?

File: lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp (hunk 2 above)
Comment (augusto2112) [reply lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:7583]: Canonical types have a pretty specific meaning I think, it's just a type without any sugar. We could change it in SwiftASTContext, but then we won't really have what is referred as a canonical type in the compiler.

File: lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp [hunk 4]
```diff
@@ -7736,6 +7741,7 @@ bool SwiftASTContext::DumpTypeValue(
   case swift::TypeKind::UnboundGeneric:
   case swift::TypeKind::BoundGenericStruct:
   case swift::TypeKind::DynamicSelf:
+    swift_can_type.dump();
```
Comment (adrian-prantl) [inline lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:7744]: This looks wrong... Shouldn't it print the result into the Stream s?

**Pull request closed without merging**